src/swap.c

```diff
@@ -195,44 +212,82 @@ static int swap_init(void) /* {{{ */
   return 0;
 } /* }}} int swap_init */
 
-static void swap_submit_usage(char const *plugin_instance, /* {{{ */
-                              gauge_t used, gauge_t free,
-                              char const *other_name, gauge_t other_value) {
-  value_list_t vl = VALUE_LIST_INIT;
-
-  vl.values = &(value_t){.gauge = NAN};
-  vl.values_len = 1;
-  sstrncpy(vl.plugin, "swap", sizeof(vl.plugin));
-  if (plugin_instance != NULL)
-    sstrncpy(vl.plugin_instance, plugin_instance, sizeof(vl.plugin_instance));
-  sstrncpy(vl.type, "swap", sizeof(vl.type));
-
-  if (values_absolute)
-    plugin_dispatch_multivalue(&vl, false, DS_TYPE_GAUGE, "used", used, "free",
-                               free, other_name, other_value, NULL);
-  if (values_percentage)
-    plugin_dispatch_multivalue(&vl, true, DS_TYPE_GAUGE, "used", used, "free",
-                               free, other_name, other_value, NULL);
-} /* }}} void swap_submit_usage */
+static void swap_append_usage3(metric_family_t *fams, char const *device,
+                               gauge_t used, gauge_t free,
+                               char const *other_name, gauge_t other) {
+  metric_family_t *fam_usage = &fams[FAM_SWAP_USAGE];
+  metric_family_t *fam_utilization = &fams[FAM_SWAP_UTILIZATION];
+
+  metric_t m = {0};
+  if (device != NULL) {
+    metric_label_set(&m, label_device, device);
+  }
+
+  bool have_other = (other_name != NULL) && !isnan(other);
+
+  if (report_usage) {
+    if (have_other) {
+      metric_family_append(fam_usage, label_state, other_name,
+                           (value_t){.gauge = other}, &m);
+    }
+
+    metric_family_append(fam_usage, label_state, state_used,
+                         (value_t){.gauge = used}, &m);
+    metric_family_append(fam_usage, label_state, state_free,
+                         (value_t){.gauge = free}, &m);
+  }
+
+  if (report_utilization) {
+    gauge_t total = used + free;
```

Comment:
Functions calling this have read/calculated total value, I think this should get that value as arg instead.

Reply:
Of all the callers for `swap_append_usage*()`, only VM_SWAPUSAGE does seem to be missing total reporting, that caller can then just pass `used+free`.

`statgrab()` version does not currently use total, but the call itself provides that: https://linux.die.net/man/3/sg_get_swap_stats

Reply:
I don't see why this is better, but done.

Reply:
Thanks, approved!

I think it's better to be explicit about it, and if OS reports total separately from used & free, use that.

While there's less likely to be a difference between total & used+free for swap, than is the case with RAM[1], multiple levels of swap, some of it e.g. compressed in RAM, others on storage with wildly different speeds, and containing overlapping data, may also make that more relevant.

[1] Memory plugin is where I really would like to see this fixed, because there I *know* used+free to be wrong value to use for total.  That being a blocker for reporting also other useful RAM usage numbers is another good reason for fixing the memory plugin, but having swap plugin reporting total properly is a good start. :-)